rfc/remove-ucan-stream.md

```diff
@@ -0,0 +1,53 @@
+# RFC: Remove the UCAN stream
+
+## Authors
+
+- [Alan Shaw](https://github.com/alanshaw), [Storacha Network](https://storacha.network/)
+
+## Introduction
+
+All invocations and receipts are put on the UCAN stream - currently an AWS Kinesis instance.
+
+The original idea of the UCAN stream was to act as an ordered log of events where (potentially) all tasks could be distributed and performed async by a decentralized network. It would allow invocations to be accepted very quickly, and tasks of any size to be completed without time limitations. Additionally the stream acted as a backup of events that could be replayed at any time to restore the current state of the system. Sadly(?) this original vision has not been realized. We instead perform most actions synchronously when they are invoked and use the stream for just a handful of tasks that relate only to post-processing.
+
+The UCAN stream exists for 2 reasons currently:
+
+1. Billing for space usage - we inspect receipts for certain invocations and add usage delta records per space in order to allow space size to be accounted for and billed.
+2. Metrics - we currently store _counts_ for certain invocations and for invocations that store or removed data, the accumulated _size_ of data added or removed.
+
+We've observed a few issues with this:
+
+* Stream processing errors lead to incorrect counts that are never rectified.
+* In the specific case of billing, we've discovered that some receipts (that are required for billing) are added onto the stream multiple times, due to the fact that they are embedded in multiple responses. Specifically the `web3.storage/blob/allocate` receipt is added onto the stream when a blob is allocated, and then again when it is included in the response for `space/blob/add`. Since [this change](https://github.com/storacha/w3infra/pull/380), we have been double counting space usage additions.
+* Async post-processing tasks i.e. tasks that don't have a direct effect on the outcome of the original invocation, can be easily lost since there is no record of them being scheduled, succeeding or failing.
+* Async tasks that depend on the receipt of a different invocation are brittle to changes in the receipt. The tasks are often "far away" from changes to the receipt they depend on and changes to the receipt may not be known to break the task until much later (or worst case at runtime in production).
```

Comment:
there's a couple things here:
- code proximity is real -- on a team this size with people floating everywhere I agree it makes sense for everyone to see everything, and asynchronous events make it natural to not be aware of logic
- "receipt not known to break a task till later" -- there are ways to mitigate perhaps via types/testing at least with certian types of queues (perhaps not Kenisis) -- but yes, it's an inherent problem.